text/0000-safer-transmute.md

```diff
@@ -0,0 +1,1722 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Terminology & Concepts
+
+### 📖 Transmutation
+**Transmutation** is the act of reinterpreting the bytes corresponding to a value of one type as if they corresponded to a different type. Concretely, we mean the behavior of this function:
+```rust
+#[inline(always)]
+unsafe fn transmute<Src, Dst>(src: Src) -> Dst
+{
+    #[repr(C)]
+    union Transmute<Src, Dst> {
+        src: ManuallyDrop<Src>,
+        dst: ManuallyDrop<Dst>,
+    }
+
+    ManuallyDrop::into_inner(Transmute { src: ManuallyDrop::new(src) }.dst)
+}
+```
+
+### 📖 Safer Transmutation
+By **safer transmutation** we mean: *what `where` bound could we add to `transmute` restricts its type parameters `Src` and `Dst` in ways that statically limit the function's misuse?* Our answer to this question will ensure that transmutations are, by default, *well-defined*, *safe* and *stable*.
+
+### 📖 Well-Definedness
+A transmutation is ***well-defined*** if the mere act of transmuting a value from one type to another is not unspecified or undefined behavior.
+
+### 📖 Safety
+A well-defined transmutation is ***safe*** if *using* the transmuted value cannot violate memory safety.
+
+### 📖 Stability
+A safe transmutation is ***stable*** if the authors of the source type and destination types have indicated that the layouts of those types is part of their libraries' stability guarantees.
+
+## Concepts in Depth
+
+***Disclaimer:** While the high-level definitions of transmutation well-definedness, safety and stability are a core component of this RFC, the detailed rules and examples in this section are **not**. We expect that the initial implementation of `TransmuteFrom` may initially be considerably less sophisticated than the examples in this section (and thus forbid valid transmutations). Nonetheless, this section explores nuanced cases of transmutation well-definedness and safety to demonstrate that the APIs we propose can grow to handle that nuance.*
+
+
+### 📖 When is a transmutation well-defined?
+[sound transmutation]: #-when-is-a-transmutation-well-defined
+A transmutation is ***well-defined*** if the mere act of transmuting a value from one type to another is not unspecified or undefined behavior.
+
+#### Well-Defined Representation
+[`u8`]: core::u8
+[`f32`]: core::f32
+
+Transmutation is ill-defined if it occurs between types with unspecified representations.
+
+Most of Rust's primitive types have specified representations. That is, the precise layout characteristics of [`u8`], [`f32`] is a documented and guaranteed aspect of those types.
+
+In contrast, most `struct` and `enum` types defined without an explicit `#[repr(C)]` attribute do ***not*** have well-specified layout characteristics.
+
+To ensure that types you've define are transmutable, you almost always (with very few exceptions) must mark them with the `#[repr(C)]` attribute.
```

Comment:
(fairly-minor point) To me I'd think there another common case: `repr(transparent)` newtypes.  That one's obviously sound, and the opt-in safety can be asymmetrical.  I'm [particularly interested](https://docs.rs/rev_slice/0.1.5/src/rev_slice/lib.rs.html#56-58) in that for `&impl !Sized`.  And I guess `repr(uN)` enums are also a plausible, more-asymmetric, case.

Should this thus be something like "a non-`repr(rust)` layout attribute"?

Reply:
> Should this thus be something like "a non-repr(rust) layout attribute"?

Even that's insufficient: there are guarantees about option-like enums, for instance.

I used the weasel phrase "almost always" because there doesn't seem to be a short answer to whether a type has a well-specified representation.